NOISSUE - Kafka support

rodneyosodo opened the pull request:
### What does this do?
Adds Kafka support. Since we are introducing a multi-broker system, we need to add support for Kafka. This allows us to use Kafka as a message broker for our system. Kafka has been chosen because it is the most popular legacy message broker. This allows us to support legacy systems that use Kafka.

### Which issue(s) does this PR fix/relate to?
Resolves #990 

### List any changes that modify/break current functionality
N/A

### Have you included tests for your changes?
Yes

### Did you document any new/modified functionality?
https://github.com/mainflux/docs/pull/157

### Notes
The publisher implementation is based on the `segmentio/kafka-go` library. Publishing messages is well supported by the library, but subscribing to topics is not. The library does not provide a way to subscribe to all topics, but only to a specific topic. This is a problem because the Mainflux platform uses a topic per channel, and the number of channels is not known in advance. The solution is to use the Zookeeper library to get a list of all topics and then subscribe to each of them. The list of topics is obtained by connecting to the Zookeeper server and reading the list of topics from the `/brokers/topics` node. The first message received from the topic can be lost if subscription happens closely followed by publishing. After the subscription, we guarantee that all messages will be received.

manuio commented on pkg/messaging/broker/pubsub_kafka.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+const (
+	// chansPrefix = "channels"
```
Remove commented code

manuio commented on pkg/messaging/broker/pubsub_nats.go:
```diff
@@ -0,0 +1,29 @@
+)
+
+const (
+	// chansPrefix = "channels"
```
Same here

manuio commented on pkg/messaging/broker/pubsub_rabbitmq.go:
```diff
@@ -0,0 +1,29 @@
+)
+
+const (
+	// chansPrefix = "channels"
```
and here

dborovcanin commented on Makefile:
```diff
@@ -16,7 +16,7 @@ TIME ?= $(shell date +%F_%T)
 define compile_service
 	CGO_ENABLED=$(CGO_ENABLED) GOOS=$(GOOS) GOARCH=$(GOARCH) GOARM=$(GOARM) \
-	go build -mod=vendor -ldflags "-s -w \
+	go build -mod=vendor -tags $(MF_BROKER_TYPE) -ldflags "-s -w \
```
Add a default value to `MF_BROKER_TYPE`, otherwise, the build will if it's empty.

drasko commented on Makefile:
```diff
@@ -111,4 +111,4 @@ rundev:
 
 run:
-	docker-compose -f docker/docker-compose.yml up
+	docker-compose -f docker/docker-compose.yml up -d
```
Remove `-d`

drasko commented on pkg/messaging/nats/pubsub.go:
```diff
@@ -15,10 +15,11 @@ import (
+const (
+	chansPrefix = "channels"
+	// SubjectAllChannels represents subject to subscribe for all the channels.
+	SubjectAllChannels = "channels.>"
```
I would construct this one as `chansPrefix + ".>"`

rodneyosodo replied:
That is what is implemented currently across all brokers. Should we change?

drasko commented on twins/README.md:
```diff
@@ -22,8 +22,8 @@ default values.
 | MF_TWINS_DB_PORT           | Database host port                                                   | 27017                 |
-| MF_THINGS_STANDALONE_EMAIL | User email for standalone mode (no gRPC communication with users)       |                |
-| MF_THINGS_STANDALONE_TOKEN | User token for standalone mode that should be passed in auth header     |                |
+| MF_THINGS_STANDALONE_EMAIL | User email for standalone mode (no gRPC communication with users)    |                       |
```
Remove tabs, use spaces

manuio commented on pkg/messaging/rabbitmq/publisher.go:
```diff
@@ -65,7 +68,7 @@ func (pub *publisher) Publish(topic string, msg messaging.Message) error {
 		amqp.Publishing{
 			Headers:     amqp.Table{},
-			ContentType: "text/plain",
+			ContentType: "application/octet-stream",
```
You can use a const here

rodneyosodo replied:
This is linked with RabbitMQ branch [here](https://github.com/mainflux/mainflux/pull/1570)

manuio commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -34,6 +34,7 @@ var (
 	errAlreadySubscribed = errors.New("already subscribed to topic")
 	errNotSubscribed     = errors.New("not subscribed")
 	errEmptyTopic        = errors.New("empty topic")
+	errEmptyID           = errors.New("empty ID")
```
If I remember well we are not using upper-case in logs

rodneyosodo replied:
This is also linked with the RabbitMQ branch let me update it [here](https://github.com/mainflux/mainflux/pull/1570)

dborovcanin commented on docker/broker/rabbitmq.yml:
```diff
@@ -0,0 +1,11 @@
+      RABBITMQ_DEFAULT_VHOST: ${MF_RABBITMQ_VHOST}
+    ports:
+      - ${MF_RABBITMQ_PORT}:${MF_RABBITMQ_PORT}
+      - ${MF_RABBITMQ_HTTP_PORT}:${MF_RABBITMQ_HTTP_PORT}
```
Please add an empty line.

dborovcanin commented on pkg/messaging/README.md:
```diff
@@ -2,8 +2,8 @@
+`Subscriber` interface defines methods used to subscribe to a message broker such as MQTT or NATS or RabbitMQ or Kafka. 
 
-`Publisher` interface defines methods used to publish messages to a message broker such as MQTT or NATS or RabbitMQ.
+`Publisher` interface defines methods used to publish messages to a message broker such as MQTT or NATS or RabbitMQ or Kafka.
```
`Publisher` interface defines methods used to publish messages to a message broker.
Currently, supported brokers are MQTT, NATS, RabbitMQ, and Kafka.

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,73 @@
+	}
+	if err := writer.WriteMessages(context.TODO(), kafkaMsg); err != nil {
+		// Sometime it take time for leader to be elected. If that is so, we retry to publish message
+		if strings.Contains(fmt.Sprint(err), "[5] Leader Not Available:") {
```
Extract this string literal to a constant if there is no other way to decide the error type. Use `err.Error()` instead of `fmt.Sprint(err)`.

dborovcanin replied:
Looks like there is already an existing exported error for this in `vendor/github.com/segmentio/kafka-go/error.goL166` so you can reuse it.

rodneyosodo replied:
I get a `runtime error: invalid memory address or nil pointer dereference [recovered]` sometimes when it tries and checks if the error is `unexpected EOF`

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,73 @@
+	if err := writer.WriteMessages(context.TODO(), kafkaMsg); err != nil {
+		// Sometime it take time for leader to be elected. If that is so, we retry to publish message
+		if strings.Contains(fmt.Sprint(err), "[5] Leader Not Available:") {
+			time.Sleep(2 * time.Second)
```
Consider adding an exponential backoff.

dborovcanin commented on pkg/messaging/kafka/pubsub.go:
```diff
@@ -0,0 +1,149 @@
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+	// Check topic
+	s, ok := ps.subscriptions[topic]
```
You can use `subs` here and later `s := subs[id]` or `sub := subs[id]` instead of `reader := subs[id]`.

dborovcanin commented on pkg/messaging/kafka/pubsub_test.go:
```diff
@@ -0,0 +1,221 @@
+		pubsub       bool //true for subscribe and false for unsubscribe
+	}{
+		{
+			desc:         "Susbcribe to a topic with an ID",
```
Fix typos `susbcribe`.

dborovcanin commented on Makefile:
```diff
@@ -123,6 +123,9 @@ ifeq ("$(MF_BROKER_TYPE)", "rabbitmq")
 	sed -i "s,MF_BROKER_URL: .*,MF_BROKER_URL: $$\{MF_NATS_URL\}," docker/docker-compose.yml
+else ifeq ("$(MF_BROKER_TYPE)", "kafka")
+	sed -i "s,file: brokers/.*.yml,file: brokers/kafka.yml," docker/docker-compose.yml
+	sed -i "s,MF_BROKER_URL: .*,MF_BROKER_URL: $$\{MF_KAFKA_URL\}," docker/docker-compose.yml
```
There is no `MF_KAFKA_URL` in `.env` file.

dborovcanin reviewed (changes requested):
Please fix CI.

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,91 @@
+
+		// Sometime it take time for leader to be elected. If that is so, we retry to publish message
+		err := writer.WriteMessages(context.TODO(), msg)
+		fmt.Println(err.Error())
```
Remove.

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,78 @@
+		Topic: subject,
+	}
+	for _, backoff := range backoffSchedule {
+		err := writer.WriteMessages(context.TODO(), kafkaMsg)
```
Since the write is set to be async, the error is unlikely to happen. It's OK to keep it async, but check what are the ways to log the possible error(s). Take a look at `ErrorLogger` field of writer config.

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,78 @@
+	if msg.Subtopic != "" {
+		subject = fmt.Sprintf("%s.%s", subject, msg.Subtopic)
+	}
+	writer := kafka.NewWriter(kafka.WriterConfig{
```
Check writer config options and try to find the best one for our use case. Looks like there are many options, backoff included (so you don't have to implement it here).

rodneyosodo replied:
They have implemented it but with no configurable parameters for the writer [here](https://github.com/mainflux/mainflux/pull/1579/files#r905379708)

dborovcanin replied:
There is a `MaxAttempts` field that can be used for that. The lib is a little bit complex, but you can try and debug the code to check how it works.

rodneyosodo replied:
It backs off up to 1s then stops

```
kafka writer: writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: error writing messages to channels.topic.engine (partition 0): kafka.(*Client).Produce: dial tcp 127.0.0.1:9092: connect: connection refused
kafka writer: backing off 100ms writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: error writing messages to channels.topic.engine (partition 0): kafka.(*Client).Produce: dial tcp 127.0.0.1:9092: connect: connection refused
kafka writer: backing off 400ms writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: error writing messages to channels.topic.engine (partition 0): kafka.(*Client).Produce: dial tcp 127.0.0.1:9092: connect: connection refused
kafka writer: backing off 900ms writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: error writing messages to channels.topic.engine (partition 0): kafka.(*Client).Produce: dial tcp 127.0.0.1:9092: connect: connection refused
kafka writer: backing off 1s writing 1 messages to channels.topic.engine (partition: 0)
```

dborovcanin replied:
That should be OK, 1s should be more than enough. In case of, say, 10 attempts - it's up to 10s. We assume that Kafka is ready to accept messages before we start and backoff is only for cases where something malfunctioning or Kafka is overloaded. How did you test this, did you wait some time after compose is up, and did you check Kafka logs to be sure it's up and ready to accept connections before you started testing?

rodneyosodo replied:
When testing it on the platform all other services throw this error message `{"level":"error","message":"Failed to connect to postgres: dial tcp 172.19.0.6:5432: connect: connection refused","ts":"2022-06-24T13:56:42.923294665Z"}`  but after that it connects. Publishing from one adapter and receiving in another works fine.

rodneyosodo commented on vendor/github.com/segmentio/kafka-go/writer.go:
```diff
@@ -0,0 +1,1272 @@
+			//   guarantees to abort, but may be better to avoid long wait times
+			//   on close.
+			//
+			delay := backoff(attempt, 100*time.Millisecond, 1*time.Second)
```
Here

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+type publisher struct {
+	conn *kafka.Conn
```
This field is not used.

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,65 @@
+	conn, err := kafka.Dial("tcp", url)
+	if err != nil {
+		return nil, err
+	}
```
It's OK to check the connection, but there is no point in keeping this in the `conn` field that is not used.

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,65 @@
+	if msg.Subtopic != "" {
+		subject = fmt.Sprintf("%s.%s", subject, msg.Subtopic)
+	}
+	writer := kafka.NewWriter(kafka.WriterConfig{
```
It looks like `kafka.NewWriter(kafka.WriterConfig)` is deprecated and the preferred way is to use the Wrteir struct directly.

dborovcanin reviewed (changes requested):
CI is failing.

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,109 @@
+		subject = fmt.Sprintf("%s.%s", subject, msg.Subtopic)
+	}
+
+	conn, err := kafka.Dial("tcp", pub.url)
```
Revert to having a connection as a `publisher` field and close it when you close the publisher, so we don't open and close the connection all the time - we keep it open.

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,109 @@
+		MaxAttempts:  100,
+		RequiredAcks: kafka.RequireAll,
+	}
+	defer writer.Close()
```
Is there a need to create a new writer always or you can always use the same one - as a field of the publisher?

dborovcanin commented on pkg/messaging/kafka/publisher.go:
```diff
@@ -0,0 +1,109 @@
+	return nil
+}
+
+func (pub *publisher) Close() error {
```
The publisher is closed only at the end of its lifetime - removing topics that way does not sound particularly useful since Publisher is meant to be long-lasting.

rodneyosodo replied:
What about cleaning up?

dborovcanin replied:
I am not sure. Is there a way to configure Kafka to delete the topic if no client has pushed to it in a while (say 1d) or no client is subscribed? If we leave topics intact, how does that affect performance?

rodneyosodo replied:
From what I have reviewed I have not seen a way to automatically delete the topic after a specific period. That option is only done with message using [delete.retention.ms](https://kafka.apache.org/documentation/#topicconfigs_delete.retention.ms). Also, the number of topics isn't all that important in and of itself, but each Kafka topic is partitioned, and the total number of partitions may have an effect on performance [[1]](https://www.instaclustr.com/blog/the-power-of-kafka-partitions-how-to-get-the-most-out-of-your-kafka-cluster/).

dborovcanin commented on pkg/messaging/kafka/pubsub.go:
```diff
@@ -0,0 +1,159 @@
+		ps.subscriptions[topic] = s
+	}
+
+	reader := kafka.NewReader(kafka.ReaderConfig{
```
What happens to the reader when there is no topic existing?

rodneyosodo replied:
If `auto.create.topics.enable=true` in your Kafka configuration file it will automatically create that topic.

dborovcanin commented on pkg/messaging/kafka/pubsub.go:
```diff
@@ -0,0 +1,159 @@
+	return nil
+}
+
+func (ps *pubsub) Close() error {
```
Simplify with something like this:
```suggestion
func (ps *pubsub) Close() error {
	ps.mu.Lock()
	defer ps.mu.Unlock()
	for _, subs := range ps.subscriptions {
		for _, s := range subs {
			if s.cancel != nil {
				if err := s.close(); err != nil {
					return err
				}
			}
		}
	}
	return nil
}
```
where `s.close` is something like this:
```go
func (s subscription) close() error {
	if err := s.cancel(); err != nil {
		return err
	}
	return s.Close()
}
```

drasko reviewed (dismissed):
LGTM

arvindh123 commented on users/postgres/users.go:
```diff
@@ -115,7 +115,7 @@ func (ur userRepository) RetrieveByEmail(ctx context.Context, email string) (use
 
 func (ur userRepository) RetrieveByID(ctx context.Context, id string) (users.User, error) {
-	q := `SELECT email, password, metadata FROM users WHERE id = $1`
+	q := `SELECT email, password, metadata FROM users WHERE state = 'active' AND id = $1`
```
This part of code is covered in this PR ? Or it should be done in separate PR , 
How we retrieve the non active user id ?

rodneyosodo replied:
No. Let me revert back

rodneyosodo committed:
Remove prev commits and rebase with master: NOISSUE-kafka-support-backup

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Fix typo

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Migrate from mainflux to magistrala

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
Updated go.mod and go.sum files with new packages and updated dependencies.

Modified Close function in Kafka messaging package and updated pubsub struct and NewPubSub function.

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
